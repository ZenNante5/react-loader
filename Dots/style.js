'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = function (style) {
  return _extends({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: '120px',
    minWidth: '120px',
    width: '100%',
    height: '100%'
  }, style);
};