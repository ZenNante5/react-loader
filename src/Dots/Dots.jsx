import React, { Component, PropTypes } from 'react'

import Dot from './Dot'
import dynStyle from './style'

const maxDots = 3
const maxOpacity = 0.6
const minOpacity = 0.1
const timeout = 1000

class Dots extends Component {
  constructor(props, context) {
    super(props, context)

    // FIXME : google help me to initialize this
    const opacities = []
    for (let i = 0; i < maxDots; ++i) opacities.push(minOpacity)

    this.state = {
      opacities,
    }

    this.intervals = []
  }

  componentDidMount() {
    for (let i = 0; i < maxDots; ++i) {
      setTimeout(() => this.runDot(i), (timeout / maxDots) * i)
    }
  }

  componentWillUnmount() {
    if (this.intervals) {
      for (const interval of this.intervals) {
        clearInterval(interval)
      }
    }
  }

  computeOpacity = (index) => {
    this.intervals.push(setInterval(() => {
      const { opacities } = this.state

      const newOpacities = [...opacities]
      newOpacities[index] = newOpacities[index] === minOpacity ? maxOpacity : minOpacity

      this.setState({
        opacities: newOpacities,
      })
    }, timeout))
  }

  runDot = (index) => this.computeOpacity(index)

  render() {
    const { style, className } = this.props
    const { opacities } = this.state

    return (
      <div style={dynStyle(style)} className={className}>
        {opacities.map((i, index) => (
          <Dot key={index} opacity={opacities[index]} />
        ))}
      </div>
    )
  }
}

Dots.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
}

export default Dots
