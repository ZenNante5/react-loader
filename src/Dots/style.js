export default (style) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  minHeight: '120px',
  minWidth: '120px',
  width: '100%',
  height: '100%',
  ...style,
})
