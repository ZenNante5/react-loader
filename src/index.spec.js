/* eslint-env mocha */
/* eslint-disable
  no-unused-expressions,
  react/jsx-filename-extension,
  import/no-extraneous-dependencies
*/

import React from 'react'
import { mount } from 'enzyme'
import { expect, spy } from 'chai'
import blanket from 'blanket' // eslint-disable-line
import loader from './index'
import Dots from './Dots'


const Component = () => <div />
const Loader = () => <div />
const getWrapped = (config, props) => {
  const Container = loader(Component, config)
  return mount(<Container {...props} />)
}


describe('react-loader', () => {
  it('should wait for a `data` props [readme]', () => {
    // Mount
    const load = spy(() => {})
    const loaded = getWrapped({ wait: ['data'] }, { load })

    // Load function is called
    // Graphic component isn't called
    // Loader should be Dots
    load.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists

    // Change `data` value
    loaded.setProps({ data: true })

    // Load function is not called twice
    // Graphic component is called
    // Loader shouldn't be printed
    load.should.have.been.called.once()
    loaded.find(Component).node.should.exists
    expect(loaded.find(Dots).node).to.be.undefined
  })

  it('should wait for a `loaded` props [default]', () => {
    // Mount
    const load = spy(() => {})
    const loaded = getWrapped(undefined, { load })

    // Load function is called
    // Graphic component isn't called
    // Loader should be Dots
    load.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists

    // Change `loaded` value
    loaded.setProps({ loaded: true })

    // Load function is not called twice
    // Graphic component is called
    // Loader shouldn't be printed
    load.should.have.been.called.once()
    loaded.find(Component).node.should.exists
    expect(loaded.find(Dots).node).to.be.undefined
  })

  it('should wait for an array of props', () => {
    // Mount
    const load = spy(() => {})
    const loaded = getWrapped({ wait: ['prop1', 'prop2'] }, { load })

    // Load function is called
    // Graphic component isn't called
    // Loader should be Dots
    load.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists

    // Change `prop1` value
    loaded.setProps({ prop1: true })

    // Load function is not called twice
    // Graphic component isn't called
    // Loader should be Dots
    load.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists

    // Change `prop3` value
    loaded.setProps({ prop3: true })

    // Load function is not called twice
    // Graphic component isn't called
    // Loader should be Dots
    load.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists

    // Change `prop2` value
    loaded.setProps({ prop2: true })

    // Load function isn't called twice
    // Graphic component is printed
    // Loader shouldn't be printed
    load.should.have.been.called.once()
    loaded.find(Component).node.should.exists
    expect(loaded.find(Dots).node).to.be.undefined
  })

  it('should handle `wait` parameter to be a function', () => {
    // Mount (false case)
    const load = spy(() => {})
    let loaded = getWrapped({ wait: () => false }, { load })

    // Load function is called
    // Graphic component isn't printed
    // Loader should be Dots
    load.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists

    // Mount (true case)
    loaded = getWrapped({ wait: () => true }, { load })

    // Load function is called
    // Graphic component is printed
    // Loader shouldn't be printed
    load.should.have.been.called.twice()
    loaded.find(Component).node.should.exists
    expect(loaded.find(Dots).node).to.be.undefined
  })

  it('should handle `wait` parameter to be a boolean', () => {
    // Mount (false case)
    const load = spy(() => {})
    let loaded = getWrapped({ wait: false }, { load })

    // Load function is called
    // Graphic component isn't called
    // Loader should be Dots
    load.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists

    // Mount (true case)
    loaded = getWrapped({ wait: true }, { load })

    // Load function is called
    // Graphic is printed
    // Loader shouldn't be printed
    load.should.have.been.called.twice()
    loaded.find(Component).node.should.exists
    expect(loaded.find(Dots).node).to.be.undefined
  })

  it('should print a different Loader component', () => {
    // Mount
    const load = spy(() => {})
    const loaded = getWrapped({ Loader, wait: ['data'] }, { load })

    // Load function is called
    // Graphic component isn't called
    // Loader should `Loader` and not `Dots`
    load.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Loader).node.should.exists
    expect(loaded.find(Dots).node).to.be.undefined

    // Change `data` value
    loaded.setProps({ data: true })

    // Load function is not called twice
    // Graphic component is called
    // Loader shouldn't be printed
    load.should.have.been.called.once()
    loaded.find(Component).node.should.exists
    expect(loaded.find(Dots).node).to.be.undefined
    expect(loaded.find(Loader).node).to.be.undefined
  })

  it('should call the `load` function parameter if present', () => {
    // Mount
    const load = spy(() => {})
    const load2 = spy(() => {})
    const loaded = getWrapped({ load: load2 }, { load })

    // Load function is called
    // Graphic component isn't called
    // Loader should be Dots
    load.should.have.been.called.once()
    load2.should.have.been.called.once()
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists
  })

  it('should handle a `null` `load` props/parameter', () => {
    // Mount
    const loaded = getWrapped()

    // Graphic component isn't called
    // Dots should be printed
    expect(loaded.find(Component).node).to.be.undefined
    loaded.find(Dots).node.should.exists
  })
})

/* eslint-enable no-unused-expressions */
