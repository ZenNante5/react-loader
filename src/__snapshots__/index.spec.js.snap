// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hoc-react-loader print parameter should not print Component -loaded props- 1`] = `"<div class=\\"tailspin\\">TailSpin</div>"`;

exports[`hoc-react-loader print parameter should not print Component -loaded props- 2`] = `"<div class=\\"tailspin\\">TailSpin</div>"`;

exports[`hoc-react-loader print parameter should not print Component -print as a value- 1`] = `"<div class=\\"tailspin\\">TailSpin</div>"`;

exports[`hoc-react-loader print parameter should not print Component -print as function- 1`] = `"<div class=\\"tailspin\\">TailSpin</div>"`;

exports[`hoc-react-loader print parameter should not print Component -print as function- 2`] = `
Array [
  Array [
    Object {
      "some": "props",
    },
    Object {},
  ],
]
`;

exports[`hoc-react-loader print parameter should print Component -empty parameters- 1`] = `"<div class=\\"component\\">Component</div>"`;

exports[`hoc-react-loader print parameter should print Component -loaded props- 1`] = `"<div class=\\"component\\">Component</div>"`;

exports[`hoc-react-loader print parameter should print Component -loaded props- 2`] = `"<div class=\\"component\\">Component</div>"`;

exports[`hoc-react-loader print parameter should print Component -print as a value- 1`] = `"<div class=\\"component\\">Component</div>"`;

exports[`hoc-react-loader print parameter should print Component -print as an array- 1`] = `"<div class=\\"tailspin\\">TailSpin</div>"`;

exports[`hoc-react-loader print parameter should print Component -print as an array- 2`] = `"<div class=\\"tailspin\\">TailSpin</div>"`;

exports[`hoc-react-loader print parameter should print Component -print as an array- 3`] = `"<div class=\\"tailspin\\">TailSpin</div>"`;

exports[`hoc-react-loader print parameter should print Component -print as an array- 4`] = `"<div class=\\"component\\">Component</div>"`;

exports[`hoc-react-loader print parameter should print Component -print as function- 1`] = `"<div class=\\"component\\">Component</div>"`;

exports[`hoc-react-loader print parameter should print Component -print as function- 2`] = `
Array [
  Array [
    Object {
      "some": "props",
    },
    Object {},
  ],
]
`;

exports[`hoc-react-loader print parameter should print Component -undefined parameters- 1`] = `"<div class=\\"component\\">Component</div>"`;
